File: packages/gh-tasks/src/commands/review.ts
import { resolveLocale, t } from '../i18n/index.ts';
import type { AppConfig } from '../lib/config.ts';
import { createClient, type GraphQLClient, resolveToken } from '../lib/github.ts';
import { type DateRange, type Period, parsePeriodFlag, rangeOf } from '../lib/period.ts';
import { ProjectError, type ProjectRef, resolveProjectRef } from '../lib/project.ts';
import {
  type ClosedIssueNode,
  GET_ORG_PROJECT_V2,
  GET_USER_PROJECT_V2,
  type GetOrgProjectV2Response,
  type GetUserProjectV2Response,
  LIST_CLOSED_ISSUES,
  LIST_MERGED_PRS,
  LIST_PROJECT_V2_ITEMS,
  type ListClosedIssuesResponse,
  type ListMergedPRsResponse,
  type ListProjectV2ItemsResponse,
  type MergedPRNode,
  type ProjectV2FieldValue,
  type ProjectV2ItemNode,
} from '../lib/queries/index.ts';
import { resolveRepo } from '../lib/repo.ts';
import { detectScope, type Scope } from '../lib/scope.ts';

export interface ReviewCommandDeps {
  client?: GraphQLClient;
  hasGitRemote?: () => boolean;
  getRemoteUrl?: () => string | null;
  stdout?: NodeJS.WritableStream;
  stderr?: NodeJS.WritableStream;
  now?: () => Date;
  config?: AppConfig;
}

const FETCH_LIMIT = 100;
const DEFAULT_PERIOD: Period = 'weekly';

/**
 * Retrospective summary for the chosen `--period`.
 *
 * - `--scope repo` (default in a Git repo): aggregates Issues `closedAt` and
 *   PRs `mergedAt` that fall in the period window.
 * - `--scope org|user`: aggregates Projects v2 items whose Status is "Done"
 *   (case-insensitive) and whose `updatedAt` falls in the period window.
 *   `updatedAt` is the proxy for "moved to Done" — Projects v2 doesn't expose
 *   per-field-change timestamps via GraphQL.
 */
export async function review(
  argv: readonly string[],
  deps: ReviewCommandDeps = {}
): Promise<number> {
  const stdout = deps.stdout ?? process.stdout;
  const stderr = deps.stderr ?? process.stderr;
  const locale = resolveLocale(argv, process.env, deps.config);

  const period = parsePeriodFlag(argv) ?? DEFAULT_PERIOD;
  const now = deps.now ? deps.now() : new Date();

  const scope = detectScope({ argv, hasGitRemote: deps.hasGitRemote, config: deps.config });

  if (scope === 'repo') {
    return await reviewRepo({ argv, deps, locale, stdout, stderr, period, now });
  }

  return await reviewProject({ argv, deps, locale, scope, stdout, stderr, period, now });
}

interface ReviewRepoContext {
  argv: readonly string[];
  deps: ReviewCommandDeps;
  locale: 'ja' | 'en';
  stdout: NodeJS.WritableStream;
  stderr: NodeJS.WritableStream;
  period: Period;
  now: Date;
}

async function reviewRepo(ctx: ReviewRepoContext): Promise<number> {
  const { argv, deps, locale, stdout, period, now } = ctx;

  const repo = resolveRepo({ argv, getRemoteUrl: deps.getRemoteUrl });
  const client = deps.client ?? createClient(resolveToken());
  const range = rangeOf(period, now);

  const [issuesData, prsData] = await Promise.all([
    client.request<ListClosedIssuesResponse>(LIST_CLOSED_ISSUES, {
      owner: repo.owner,
      name: repo.name,
      first: FETCH_LIMIT,
    }),
    client.request<ListMergedPRsResponse>(LIST_MERGED_PRS, {
      owner: repo.owner,
      name: repo.name,
      first: FETCH_LIMIT,
    }),
  ]);

  const closedIssues =
    issuesData.repository?.issues.nodes.filter((n) => withinRange(n.closedAt, range)) ?? [];
  const mergedPRs =
    prsData.repository?.pullRequests.nodes.filter((n) => withinRange(n.mergedAt, range)) ?? [];

  stdout.write(renderRepoMarkdown({ period, range, closedIssues, mergedPRs, locale }));
  return 0;
}

interface ReviewProjectContext {
  argv: readonly string[];
  deps: ReviewCommandDeps;
  locale: 'ja' | 'en';
  scope: Exclude<Scope, 'repo'>;
  stdout: NodeJS.WritableStream;
  stderr: NodeJS.WritableStream;
  period: Period;
  now: Date;
}

async function reviewProject(ctx: ReviewProjectContext): Promise<number> {
  const { argv, deps, locale, scope, stdout, stderr, period, now } = ctx;

  let projectRef: ProjectRef;
  try {
    projectRef = resolveProjectRef({ scope, argv, config: deps.config });
  } catch (err) {
    if (err instanceof ProjectError) {
      stderr.write(`${err.message}\n`);
      return 2;
    }
    throw err;
  }

  const client = deps.client ?? createClient(resolveToken());
  const projectId = await resolveProjectNodeId({ client, scope, projectRef });
  if (!projectId) {
    stderr.write(
      `project not found: ${projectRef.owner}/${projectRef.number} (--scope ${scope})\n`
    );
    return 1;
  }

  const data = await client.request<ListProjectV2ItemsResponse>(LIST_PROJECT_V2_ITEMS, {
    projectId,
    first: FETCH_LIMIT,
  });
  if (!data.node) {
    stderr.write(
      `project not found: ${projectRef.owner}/${projectRef.number} (--scope ${scope})\n`
    );
    return 1;
  }

  const range = rangeOf(period, now);
  const completed = data.node.items.nodes.filter(
    (item) => withinRange(item.updatedAt, range) && isDone(item)
  );

  stdout.write(renderProjectMarkdown({ period, range, completed, locale }));
  return 0;
}

function withinRange(iso: string | null | undefined, range: DateRange): boolean {
  if (!iso) return false;
  const ms = new Date(iso).getTime();
  if (!Number.isFinite(ms)) return false;
  return ms >= range.start.getTime() && ms < range.end.getTime();
}

/** Status is "Done" (case-insensitive). Status field is conventionally named "Status". */
function isDone(item: ProjectV2ItemNode): boolean {
  const status = findStatus(item.fieldValues.nodes);
  if (status === null) return false;
  return status.toLowerCase() === 'done';
}

function findStatus(values: readonly ProjectV2FieldValue[]): string | null {
  for (const v of values) {
    if (
      v.__typename === 'ProjectV2ItemFieldSingleSelectValue' &&
      v.field.name.toLowerCase() === 'status'
    ) {
      return v.name;
    }
  }
  return null;
}

interface RenderRepoInput {
  period: Period;
  range: DateRange;
  closedIssues: ClosedIssueNode[];
  mergedPRs: MergedPRNode[];
  locale: 'ja' | 'en';
}

function renderRepoMarkdown(input: RenderRepoInput): string {
  const { period, range, closedIssues, mergedPRs, locale } = input;
  const startStr = range.start.toISOString().slice(0, 10);
  const endStr = range.end.toISOString().slice(0, 10);
  const lines: string[] = [];
  lines.push(`# ${t(locale, 'review.heading')} (${period})`);
  lines.push(`${startStr} → ${endStr}`);
  lines.push('');
  lines.push(`## ${t(locale, 'review.closedIssues')} (${closedIssues.length})`);
  if (closedIssues.length === 0) {
    lines.push(`- ${t(locale, 'review.none')}`);
  } else {
    for (const i of closedIssues) {
      lines.push(`- #${i.number} ${i.title} (${i.url})`);
    }
  }
  lines.push('');
  lines.push(`## ${t(locale, 'review.mergedPRs')} (${mergedPRs.length})`);
  if (mergedPRs.length === 0) {
    lines.push(`- ${t(locale, 'review.none')}`);
  } else {
    for (const p of mergedPRs) {
      lines.push(`- #${p.number} ${p.title} (${p.url})`);
    }
  }
  lines.push('');
  return lines.join('\n');
}

interface RenderProjectInput {
  period: Period;
  range: DateRange;
  completed: ProjectV2ItemNode[];
  locale: 'ja' | 'en';
}

function renderProjectMarkdown(input: RenderProjectInput): string {
  const { period, range, completed, locale } = input;
  const startStr = range.start.toISOString().slice(0, 10);
  const endStr = range.end.toISOString().slice(0, 10);
  const lines: string[] = [];
  lines.push(`# ${t(locale, 'review.heading')} (${period})`);
  lines.push(`${startStr} → ${endStr}`);
  lines.push('');
  lines.push(`## ${t(locale, 'review.completedProjectItems')} (${completed.length})`);
  if (completed.length === 0) {
    lines.push(`- ${t(locale, 'review.empty.project')}`);
  } else {
    for (const item of completed) {
      lines.push(`- ${formatItemLine(item)}`);
    }
  }
  lines.push('');
  return lines.join('\n');
}

function formatItemLine(item: ProjectV2ItemNode): string {
  const c = item.content;
  if (!c) return '(no content)';
  if (c.__typename === 'Issue' || c.__typename === 'PullRequest') {
    const prefix = c.__typename === 'PullRequest' ? 'PR' : '';
    return `${prefix}#${c.number} ${c.title} (${c.url})`;
  }
  return `(draft) ${c.title}`;
}

interface ResolveProjectNodeIdOptions {
  client: GraphQLClient;
  scope: Exclude<Scope, 'repo'>;
  projectRef: ProjectRef;
}

async function resolveProjectNodeId(opts: ResolveProjectNodeIdOptions): Promise<string | null> {
  const { client, scope, projectRef } = opts;
  if (scope === 'org') {
    const data = await client.request<GetOrgProjectV2Response>(GET_ORG_PROJECT_V2, {
      login: projectRef.owner,
      number: projectRef.number,
    });
    return data.organization?.projectV2?.id ?? null;
  }
  const data = await client.request<GetUserProjectV2Response>(GET_USER_PROJECT_V2, {
    login: projectRef.owner,
    number: projectRef.number,
  });
  return data.user?.projectV2?.id ?? null;
}

File: cmd/review.go
package cmd

import (
	"context"
	"fmt"
	"time"

	"github.com/spf13/cobra"

	"github.com/ozzy-labs/gh-tasks/internal/github/queries"
	"github.com/ozzy-labs/gh-tasks/internal/period"
	"github.com/ozzy-labs/gh-tasks/internal/project"
	"github.com/ozzy-labs/gh-tasks/internal/projectitem"
	"github.com/ozzy-labs/gh-tasks/internal/repo"
	"github.com/ozzy-labs/gh-tasks/internal/scope"
)

const reviewFetchLimit = 100

func newReviewCmd(deps Deps) *cobra.Command {
	c := &cobra.Command{
		Use:   "review",
		Short: "Generate retrospective summary",
		RunE: func(c *cobra.Command, _ []string) error {
			return runReview(c.Context(), c, deps)
		},
	}
	c.Flags().String("period", "weekly", "aggregation window: daily | weekly | sprint")
	return c
}

func runReview(ctx context.Context, c *cobra.Command, deps Deps) error {
	r, err := deps.Resolve()
	if err != nil {
		return localizedError(c, r, err)
	}
	pflag, _ := c.Flags().GetString("period")
	p, err := period.Parse(pflag)
	if err != nil {
		return localizedError(c, r, err)
	}
	rng := period.Of(p, period.Options{Getenv: deps.Env, Now: deps.Now()})
	sc, err := scope.Detect(scope.DetectOptions{
		Argv:         deps.Argv,
		HasGitRemote: deps.HasGitRemote,
		DefaultScope: r.Config.DefaultScope,
	})
	if err != nil {
		return localizedError(c, r, err)
	}
	if sc == scope.Repo {
		return runReviewRepo(ctx, c, deps, r, p, rng)
	}
	return runReviewProject(ctx, c, deps, r, sc, p, rng)
}

func runReviewRepo(ctx context.Context, c *cobra.Command, deps Deps, r Resolved, p period.Period, rng period.Range) error {
	id, err := repo.Resolve(repo.ResolveOptions{Argv: deps.Argv, GetRemoteURL: deps.GetRemoteURL})
	if err != nil {
		return localizedError(c, r, err)
	}
	clients, err := deps.NewClients()
	if err != nil {
		return localizedError(c, r, err)
	}
	gqlClient := clients.AsGenqlientClient()
	closedResp, err := queries.ListClosedIssues(ctx, gqlClient, id.Owner, id.Name, reviewFetchLimit)
	if err != nil {
		return fmt.Errorf("list closed issues: %w", err)
	}
	prsResp, err := queries.ListMergedPRs(ctx, gqlClient, id.Owner, id.Name, reviewFetchLimit)
	if err != nil {
		return fmt.Errorf("list merged PRs: %w", err)
	}
	type closedItem struct {
		Number int
		Title  string
		URL    string
	}
	closed := []closedItem{}
	if closedResp.Repository != nil {
		for _, n := range closedResp.Repository.Issues.Nodes {
			closedAt := ""
			if n.ClosedAt != nil {
				closedAt = *n.ClosedAt
			}
			if withinPeriodRange(closedAt, rng) {
				closed = append(closed, closedItem{Number: n.Number, Title: n.Title, URL: n.Url})
			}
		}
	}
	type mergedItem struct {
		Number int
		Title  string
		URL    string
	}
	merged := []mergedItem{}
	if prsResp.Repository != nil {
		for _, n := range prsResp.Repository.PullRequests.Nodes {
			mergedAt := ""
			if n.MergedAt != nil {
				mergedAt = *n.MergedAt
			}
			if withinPeriodRange(mergedAt, rng) {
				merged = append(merged, mergedItem{Number: n.Number, Title: n.Title, URL: n.Url})
			}
		}
	}
	out := c.OutOrStdout()
	fmt.Fprintf(out, "# %s (%s)\n", r.T("review.heading"), r.T("review.period."+string(p)))
	fmt.Fprintf(out, "%s → %s\n\n", rng.Start.Format("2006-01-02"), rng.End.Format("2006-01-02"))
	fmt.Fprintf(out, "## %s (%d)\n", r.T("review.closedIssues"), len(closed))
	if len(closed) == 0 {
		fmt.Fprintf(out, "- %s\n", r.T("review.none"))
	} else {
		for _, i := range closed {
			fmt.Fprintf(out, "- #%d %s (%s)\n", i.Number, i.Title, i.URL)
		}
	}
	fmt.Fprintf(out, "\n## %s (%d)\n", r.T("review.mergedPRs"), len(merged))
	if len(merged) == 0 {
		fmt.Fprintf(out, "- %s\n", r.T("review.none"))
	} else {
		for _, p := range merged {
			fmt.Fprintf(out, "- #%d %s (%s)\n", p.Number, p.Title, p.URL)
		}
	}
	fmt.Fprintln(out)
	return nil
}

func runReviewProject(ctx context.Context, c *cobra.Command, deps Deps, r Resolved, sc scope.Scope, p period.Period, rng period.Range) error {
	pref, err := project.Resolve(project.ResolveOptions{
		Scope:       sc,
		Argv:        deps.Argv,
		OrgProject:  r.Config.OrgProject,
		UserProject: r.Config.UserProject,
	})
	if err != nil {
		return localizedError(c, r, err)
	}
	clients, err := deps.NewClients()
	if err != nil {
		return localizedError(c, r, err)
	}
	pid, err := projectitem.ResolveProjectNodeID(ctx, clients.GraphQL, sc, pref)
	if err != nil {
		return err
	}
	if pid == "" {
		fmt.Fprintln(c.ErrOrStderr(), r.T("error.project.notFound", "owner", pref.Owner, "number", pref.Number, "scope", sc))
		return ErrSilentRuntime
	}
	var resp queries.ListProjectV2ItemsResponse
	if err := clients.GraphQL.Do(ctx, queries.ListProjectV2Items, map[string]any{
		"projectId": pid, "first": reviewFetchLimit,
	}, &resp); err != nil {
		return fmt.Errorf("list project items: %w", err)
	}
	if resp.Node == nil {
		fmt.Fprintln(c.ErrOrStderr(), r.T("error.project.notFound", "owner", pref.Owner, "number", pref.Number, "scope", sc))
		return ErrSilentRuntime
	}
	completed := []queries.ProjectV2ItemNode{}
	for _, item := range resp.Node.Items.Nodes {
		if withinPeriodRange(item.UpdatedAt, rng) && isItemDone(item) {
			completed = append(completed, item)
		}
	}
	out := c.OutOrStdout()
	fmt.Fprintf(out, "# %s (%s)\n", r.T("review.heading"), r.T("review.period."+string(p)))
	fmt.Fprintf(out, "%s → %s\n\n", rng.Start.Format("2006-01-02"), rng.End.Format("2006-01-02"))
	fmt.Fprintf(out, "## %s (%d)\n", r.T("review.completedProjectItems"), len(completed))
	if len(completed) == 0 {
		fmt.Fprintf(out, "- %s\n", r.T("review.empty.project"))
	} else {
		for _, item := range completed {
			fmt.Fprintf(out, "- %s\n", projectitem.FormatItemLineCompact(item))
		}
	}
	fmt.Fprintln(out)
	return nil
}

func withinPeriodRange(iso string, rng period.Range) bool {
	if iso == "" {
		return false
	}
	t, err := time.Parse(time.RFC3339, iso)
	if err != nil {
		return false
	}
	return (t.Equal(rng.Start) || t.After(rng.Start)) && t.Before(rng.End)
}
